### ISSUE TITLE
Support for virtual threads (Java 21)

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-java-instrumentation#9534]
Virtual threads are out of preview and included in new Java 21 (LTS) release (https://openjdk.org/jeps/444).

Notes/references from weekly Java meeting:

* https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything 
* https://github.com/elastic/apm-agent-java/issues/3059
  * ThreadLocals are still supported, but can be more memory intensive for things like caches of non-thread-safe objects (e.g. SimpleDateFormat) now that there can be huge numbers of virtual threads
* There is a new executor service which needs to be instrumented

**Comment by JonasKunz** [comment]
I looked for all usages of `ThreadLocal` in the API, SDK and Agent and tried to classify for which usages action might be required for handling virtual threads nicely.

## Usages might needing work

API: [TemporaryBuffers](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/api/all/src/main/java/io/opentelemetry/api/internal/TemporaryBuffers.java#L20)

* All except one usages acquire a very small buffer (<64 chars, so <128 byte)
  * [SpanId parsing/generation](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/api/all/src/main/java/io/opentelemetry/api/trace/SpanId.java#L79) (16 chars)
  * [TraceId parsing/generation](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/api/all/src/main/java/io/opentelemetry/api/trace/TraceId.java#L83) (32 chars)
  * [W3CTraceContextPropagator](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/api/all/src/main/java/io/opentelemetry/api/trace/propagation/W3CTraceContextPropagator.java#L106) (55 chars)
* The only "large" buffer is allocated by [PercentEscaper](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/api/all/src/main/java/io/opentelemetry/api/internal/PercentEscaper.java#L148) (1000 chars, so 2000 bytes)
* Proposed Solution:
  *  Stay with thread locals for <64 chars for performance
  *  Use a char[] object pool for `PercentEscaper`. The two atomic operations for pool management shouldn't have too much overhead in relation to the percent escaping logic. It is only on a slow path anyway
  *  Not too urgent for now, as this code path is only entered when actually requiring encoding (e.g. special characters in propagated baggage)

SDK: JdkHttpSender

*  [`ThreadLocal<ByteBufferPool>`](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java#L47) is very heavy-weight
*  Should be okay though, because this code path is only executed for a private thread-pool with five threads
*  We would only need to adapt this if we plan to replace that thread-pool with virtual threads at some time, because then the pooling will become ineffective (but no memory leak due to prolonged lifetime)

SDK: ProtoSerializer

* Uses [`ThreadLocal<OutputStreamEncoder>`](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/CodedOutputStream.java#L85) with 50kb buffer
* [`ThreadLocal<Map<String, byte[]>> THREAD_LOCAL_ID_CACHE`](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/ProtoSerializer.java#L25) is fine because it is cleared on `ProtoSerializer.close()`
* Assumption: ProtoSerializer is only called from very few exporter threads, correct? Then the `ThreadLocal`s are fine.
  * If we plan on ever replacing those Threads with virtual ones the pooling would become ineffective (but no memory leak due to prolonged lifetime)

For the agent I think no changes are needed, all usages seem fine to me!

## Usages looking fine to me

I checked the following usages across API, SDK and Agent but classified them as non-problematic for virtual threads:

API:

* [ThreadLocalContextStorage](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/context/src/main/java/io/opentelemetry/context/ThreadLocalContextStorage.java#L17): Cleared when no context is active

SDK:

* WeakConcurrentMap: [`ThreadLocal<LookupKey<?>>`](https://github.com/open-telemetry/opentelemetry-java/blob/5ccc7fccb3fd0bdfa902bfea7bcee43384cdcf3e/context/src/main/java/io/opentelemetry/context/internal/shaded/WeakConcurrentMap.java#L62)
  * Expected to be used very frequently, so an object pool would be bad for performance imo
  * Object is very small, so okay to have as "deadweight" even for virtual threads

Agent:

* Usages of `ThreadLocal` with a short lifespan (e.g. `ThreadLocal` is cleared after an instrumented method ends):
  * Kafka Streams [StateHolder](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/kafka/kafka-streams-0.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/StateHolder.java#L13)
  * [JedisRequestContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/jedis/jedis-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jedis/JedisRequestContext.java#L12)
  * Kubernetes client [CurrentState](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/kubernetes-client-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kubernetesclient/CurrentState.java#L18C43-L18C79)
  * Liberty [ThreadLocalContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/liberty/liberty-20.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/liberty/ThreadLocalContext.java#L16)
  * RestEasy [OtelRequestContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/quarkus-resteasy-reactive/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/quarkus/resteasy/reactive/OtelRequestContext.java#L11)
  * [RabbitMQ `ThreadLocal<Context>`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/rabbitmq-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rabbitmq/RabbitCommandInstrumentation.java#L44)
  * RMI [ThreadLocalContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/rmi/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/rmi/ThreadLocalContext.java#L11)
  * [AgentBuilderUtil.TransformContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/javaagent-tooling/src/main/java/net/bytebuddy/agent/builder/AgentBuilderUtil.java#L303)
* Usages of `ThreadLocal` which are not cleared before thread death, but the objects are very small:
  * [KafkaClientsConsumerProcessTracing](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/kafka/kafka-clients/kafka-clients-0.11/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/kafka/KafkaClientsConsumerProcessTracing.java#L13)
  * [Spring TracingChannelInterceptor `ThreadLocal<Map<MessageChannel, ContextAndScope>>`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/spring/spring-integration-4.1/library/src/main/java/io/opentelemetry/instrumentation/spring/integration/v4_1/TracingChannelInterceptor.java#L32)
    * The map remains, but it is cleared
  * [SpringSchedulingTaskTracing](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/spring/spring-scheduling-3.1/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/spring/SpringSchedulingTaskTracing.java#L10)
  * [Vert.X `ThreadLocal<SqlConnectOptions>`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/instrumentation/vertx/vertx-sql-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/sql/VertxSqlClientSingletons.java#L33C24-L33C24)
    * Not cleared after usage, therefore the lifetime of SqlConnectOptions is prolonged. However, I would assume that the number of actual SqlConnectOptions instances is low enough so that this doesn't matter.
  * [InClassLoaderMatcher `ThreadLocal<boolean[]>`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/internal/InClassLoaderMatcher.java#L15)
  * [DefineClassHandler `ThreadLocal<DefineClassContextImpl>`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d4f7135d4dee71740f9ab0716b969bf88f6aed8b/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/DefineClassHandler.java#L17)

# TL;DR

 * All `ThreadLocal` usages in the agent are fine
 * Under the assumption that exporters are not using virtual threads we are good in the SDK
 * For the API we only need to look at the `TemporaryBuffers` usage by `PercentEscaper`.
   *  Imo it would be worth replacing this usage with an object-pool instead
   *  Not too urgent, as excessive memory usage would be an edge case even with lots of virtual threads

Some Feedback on whether you agree / disagree with would be great!

**Comment by trask** [comment]
Re-closing this as "completed". Please open a bug report if there are any issues with the virtual thread support.
